modules/bpi/bpi.syndicate.inc

```diff
@@ -682,17 +694,56 @@ function bpi_available_map() {
  * @param int $bpi_id
  *   Content ID, as as stored in BPI service.
  */
-function bpi_syndicate_action($bpi_id) {
-  $bpi_ctype = variable_get('bpi_content_type', '');
-
-  if (!empty($bpi_ctype)) {
-    drupal_goto(str_replace('_', '-', 'node/add/' . $bpi_ctype), array('query' => array('bpi_id' => $bpi_id)));
+function bpi_syndicate_action($bpi_type, $bpi_id) {
+  $mapping = bpi_get_mapping($bpi_type);
+
+  if (!empty($mapping) && !empty($mapping['bpi_content_type'])) {
+    $hook = 'bpi_syndicate_action_url';
+    foreach (module_implements($hook) as $module) {
+      $url = module_invoke($module, $hook, $bpi_type, $bpi_id, $mapping);
+      if (!empty($url)) {
+        header('Location: ' . $url, TRUE, 301);
+        drupal_exit($url);
+      }
+    }
+    drupal_set_message(t('Cannot syndicate @type', array('@type' => $bpi_type)), 'warning');
+    drupal_goto('admin/bpi');
   }
   else {
+    drupal_set_message(t('Cannot find BPI mapping for @type', array('@type' => $bpi_type)), 'warning');
     drupal_goto('admin/bpi');
   }
 }
 
+/**
+ * Get url for syndicating a BPI node.
+ *
+ * All modules implementing BPI syndication will be asked for the url used to
+ * syndicate the node.
+ *
+ * @param string $bpi_type
+ *   The bpi type.
+ * @param string $bpi_id
+ *   The bpi node id.
+ *
+ * @return string|NULL
+ *   The syndicate url.
+ */
+function bpi_get_syndicate_url($bpi_type, $bpi_id) {
+  $mapping = bpi_get_mapping($bpi_type);
+  if (!empty($mapping)) {
+    $hook = 'bpi_syndicate_action_url';
+    foreach (module_implements($hook) as $module) {
+      $url = module_invoke($module, $hook, $bpi_type, $bpi_id, $mapping);
+      if (!empty($url)) {
```

Comment:
This pattern where we only wants the first hook implementation to run seems to be a common theme. How about implementing a `bpi_module_invoke_one()` function for this.

If it is/becomes a reoccuring pattern throughout the platform then we may move it to a more generic namespace.

Reply:
Good idea!